### PR TITLE
DSA 186-4 Key Generation, raw params and key import/export

### PR DESCRIPTION
**Pull request by cconlon**
This PR makes wc_MakeDsaKey() compliant with FIPS 186-4 key generation, using "Key Pair Generation Using Extra Random Bits" from section B.1.1.

This also adds functions for import/export of raw DSA parameters, and an export function for raw DSA key members.  These will be needed for CAVP validation tests.

```
int wc_DsaImportParamsRaw(DsaKey* dsa, const char* p, const char* q, const char* g);
int wc_DsaExportParamsRaw(DsaKey* dsa, byte* p, word32* pSz, byte* q, word32* qSz,
                          byte* g, word32* gSz);
int wc_DsaExportKeyRaw(DsaKey* dsa, byte* x, word32* xSz, byte* y, word32* ySz);
```

Note: this does not include 186-4 parameter generation.

### CODE REVIEW

File: wolfcrypt/src/dsa.c [hunk 1]
```diff
@@ -97,68 +97,133 @@ void wc_FreeDsaKey(DsaKey* key)
     mp_clear(&key->p);
 }
 
+
+/* validate that (L,N) match allowed sizes from FIPS 186-4, Section 4.2.
+ * l - represents L, the size of p in bits
```
Comment (toddouska) [inline wolfcrypt/src/dsa.c]: lowercase l on it's own is a poor variable name, looks like I and 1 depending on font.

File: wolfcrypt/src/dsa.c [hunk 2]
```diff
@@ -97,68 +97,133 @@ void wc_FreeDsaKey(DsaKey* key)
...
     if (rng == NULL || dsa == NULL)
         return BAD_FUNC_ARG;
 
-    qsize = mp_unsigned_bin_size(&dsa->q);
-    if (qsize == 0)
+    qSz = mp_unsigned_bin_size(&dsa->q);
+    pSz = mp_unsigned_bin_size(&dsa->p);
+
+    /* verify (L,N) pair bit lengths */
+    if (CheckDsaLN(pSz * 8, qSz * 8) != 0)
```
Comment (toddouska) [inline wolfcrypt/src/dsa.c]: Shouldn't this be WOLFSSL_BIT_SIZE instead of 8?

File: wolfcrypt/src/dsa.c [hunk 3]
```diff
@@ -97,68 +97,133 @@ void wc_FreeDsaKey(DsaKey* key)
...
         return BAD_FUNC_ARG;
 
-    /* allocate ram */
-    buf = (unsigned char *)XMALLOC(qsize, dsa->heap,
-                                   DYNAMIC_TYPE_TMP_BUFFER);
-    if (buf == NULL)
+    /* generate extra 64 bits so that bias from mod function is negligible */
+    cSz = qSz + 8;
```
Comment (toddouska) [inline wolfcrypt/src/dsa.c]: Isn't 64 / WOLFSSL_BIT_SIZE more clear since your comment says 64 and the code says 8.

File: wolfcrypt/src/dsa.c [hunk 4]
```diff
@@ -361,6 +426,157 @@ int wc_MakeDsaParameters(WC_RNG *rng, int modulus_size, DsaKey *dsa)
...
+
+    if (dsa == NULL || p == NULL || pSz == NULL || q == NULL ||
+        qSz == NULL || g == NULL || gSz == NULL)
+        return BAD_FUNC_ARG;
+
+    /* export p */
+    tmp = mp_unsigned_bin_size(&dsa->p);
+    if (*pSz < tmp) {
+        WOLFSSL_MSG("Output buffer for DSA p parameter too small");
+        return BUFFER_E;
```
Comment (toddouska) [inline wolfcrypt/src/dsa.c]: How is the caller going to know how big pSz should be when they call again?  Same for q and g.

File: wolfcrypt/src/dsa.c [hunk 5]
```diff
@@ -361,6 +426,157 @@ int wc_MakeDsaParameters(WC_RNG *rng, int modulus_size, DsaKey *dsa)
...
+    word32 tmp;
+
+    if (dsa == NULL || x == NULL || xSz == NULL || y == NULL || ySz == NULL)
+        return BAD_FUNC_ARG;
+
+    /* export x */
+    tmp = mp_unsigned_bin_size(&dsa->x);
+    if (*xSz < tmp) {
+        WOLFSSL_MSG("Output buffer for DSA private key (x) too small");
+        return BUFFER_E;
```
Comment (toddouska) [inline wolfcrypt/src/dsa.c]: We should help the caller by setting the needed size here and for y.

**Comment by cconlon** [comment]
Thanks for the feedback, fixes pushed

**Pull request merged**